pyi.py

```diff
@@ -53,58 +53,71 @@ class TypeVarInfo(NamedTuple):
-    "typing.Deque": "collections.deque",
-    "typing.DefaultDict": "collections.defaultdict",
-    "typing.ChainMap": "collections.ChainMap",
+    "typing.Counter": ("collections.Counter", _COUNTER_SLICE),
+    "typing.Deque": ("collections.deque", _SEQUENCE_SLICE),
+    "typing.DefaultDict": ("collections.defaultdict", _MAPPING_SLICE),
+    "typing.ChainMap": ("collections.ChainMap", _MAPPING_SLICE),
     # typing aliases for builtins
-    "typing.Dict": "builtins.dict",
-    "typing.FrozenSet": "builtins.frozenset",
-    "typing.List": "builtins.list",
-    "typing.Set": "builtins.set",
-    "typing.Tuple": "builtins.tuple",
-    "typing.Type": "builtins.type",
+    "typing.Dict": ("dict", _MAPPING_SLICE),
+    "typing.FrozenSet": ("frozenset", _SET_SLICE),
+    "typing.List": ("list", _SEQUENCE_SLICE),
+    "typing.Set": ("set", _SET_SLICE),
+    "typing.Tuple": ("tuple", "Foo, Bar"),
+    "typing.Type": ("type", _TYPE_SLICE),
     # One typing alias for contextlib
-    "typing.AsyncContextManager": "contextlib.AbstractAsyncContextManager",
+    "typing.AsyncContextManager": (
+        "contextlib.AbstractAsyncContextManager",
+        _CONTEXTLIB_SLICE,
+    ),
     # typing_extensions aliases for collections
-    "typing_extensions.Counter": "collections.Counter",
-    "typing_extensions.Deque": "collections.deque",
-    "typing_extensions.DefaultDict": "collections.defaultdict",
-    "typing_extensions.ChainMap": "collections.ChainMap",
+    "typing_extensions.Counter": ("collections.Counter", _COUNTER_SLICE),
+    "typing_extensions.Deque": ("collections.deque", _SEQUENCE_SLICE),
+    "typing_extensions.DefaultDict": ("collections.defaultdict", _MAPPING_SLICE),
+    "typing_extensions.ChainMap": ("collections.ChainMap", _MAPPING_SLICE),
     # One typing_extensions alias for a builtin
-    "typing_extensions.Type": "builtins.type",
+    "typing_extensions.Type": ("type", _TYPE_SLICE),
     # one typing_extensions alias for contextlib
-    "typing_extensions.AsyncContextManager": "contextlib.AbstractAsyncContextManager",
+    "typing_extensions.AsyncContextManager": (
+        "contextlib.AbstractAsyncContextManager",
+        _CONTEXTLIB_SLICE,
+    ),
 }
 
 # typing_extensions.ContextManager is omitted from the Y023 and Y027 collections - special-cased
-_BAD_Y023_IMPORTS = frozenset(
-    {
-        # collections.abc aliases
-        "Awaitable",
-        "Coroutine",
-        "AsyncIterable",
-        "AsyncIterator",
-        "AsyncGenerator",
-        # typing aliases
-        "Protocol",
-        "runtime_checkable",
-        "ClassVar",
-        "NewType",
-        "overload",
-        "Text",
-        "NoReturn",
-    }
-)
+# We use `None` to signify that the object shouldn't  be parameterised.
+_BAD_Y023_IMPORTS = {
+    # collections.abc aliases
+    "Awaitable": "T",
+    "Coroutine": "YieldType, SendType, ReturnType",
+    "AsyncIterable": "T",
+    "AsyncIterator": "T",
+    "AsyncGenerator": "YieldType, SendType",
```

Comment:
Does this really not have a return type parameter like `Coroutine` and `Generator` have?

Reply:
Yeah, it doesn't. From the [docs](https://docs.python.org/3/library/typing.html#typing.AsyncGenerator):

> Unlike normal generators, async generators cannot return a value, so there is no ReturnType type parameter.